First stab at GroupContextExtensionsDiff (#2)
bifurcation/mls-appsync: draft-barnes-mls-appsync.md

---

> title: "Using Messaging Layer Security to Modify GroupContext Extensions"

Reviewer: ```suggestion
title: "Efficient Updates to Messaging Layer Security GroupContext Extensions"
```

--- abstract

One feature that the Messaging Layer Security (MLS) protocol provides is that it

> allows the members of a group to confirm that they agree on certain data.

Reviewer: I would rephrase as 

> The Messaging Layer Security (MLS) protocol allows the members of the group to agree on a set of GroupContext extensions.

MLS includes a mechanism to modify this data (the GroupContext) all at once,

> but not to modify it individually. In this document, we define a mechanism

Reviewer: ```suggestion
MLS includes a mechanism to do a wholesale replacements of all GroupContext extensions,
but not to modify individual extensions. In this document, we define a mechanism
```

Unfortunately this state can only be modified using the
GroupContextExtensions Proposal, which needs to include the entire
GroupContext. This may include dozens of individual extensions and may be
quite large. MLS clients should be able to modify orthogonal aspects of the
GroupContext in separate Proposals and should not need to send large
amounts of data for a small change.

Here, we define a new MLS proposal type GroupContextExtensionsDiff. This proposal type allows modification on a per extension basis. It also allows GroupContext extensions to define simple diff formats (for example, to add, remove, and update elements in a list or items in a map).

# Conventions and Definitions

{::boilerplate bcp14-tagged}

This document uses many terms from {{!RFC9420}}. Some of these have
names which are similar or may be confusing: GroupContext, GroupContext
extension, ExtensionType, and GroupContextExtensions proposal.

# GroupContextDiff

This document defines a new GroupContextExtensionsDiff proposal. It
is not a Safe Extension as defined in {{Section 2 of
!I-D.ietf-mls-extensions}}, because it can modify any GroupContext
extension, including those defined in {{!RFC9420}}.

> The GroupContextExtensionsDiff proposal does not require an UpdatePath,

Author: ```suggestion
The GroupContextExtensionsDiff proposal requires an UpdatePath,
```

cause the second inserted element to be *before* the first inserted element.

Attempting to insert or replace at a non-existing index renders the diff
and its proposal invalid.

# Security Considerations

The mechanism defined in this document provides strong authenticity, integrity,
and change control properties to the state information it manages.
No unauthorized parties can make changes to the GroupContext, and the
identity of the group member making each change is authenticated.

The data synchronized via this mechanism may or may not be confidential to
the group, depending on whether the GroupContextExtensionsDiff proposal is
sent as
an MLS PublicMessage or PrivateMessage.  As with application data, applications
should generally prefer the use of Private Message.  There may be cases,
however, where it is useful for intermediaries to inspect application state
updates, e.g., to enforce policy.

# IANA Considerations

## GroupContextExtensionDiff Proposal

The `gce_diff` MLS Proposal Type is used to update Group Context Extensions
in a group more efficiently than using a `group_context_extensions` proposal

> type. The `gce_update` type is updating rather than replacing the extensions.

Author: ```suggestion
type. The `gce_diff` type is updating rather than replacing the extensions.
```

* Value: 0x000d
* Name: gce_diff
* Recommended: Y
* External: Y

> * Path Required: N

Author: ```suggestion
* Path Required: Y
```

metadata in order to confirm that the members of an MLS group agree on
application state as well as MLS metadata.

Unfortunately this state can only be modified using the
GroupContextExtensions Proposal, which needs to include the entire
GroupContext. This may include dozens of individual extensions and may be
quite large. MLS clients should be able to modify orthogonal aspects of the
GroupContext in separate Proposals and should not need to send large
amounts of data for a small change.

Here, we define a new MLS proposal type GroupContextExtensionsDiff. This proposal type allows modification on a per extension basis. It also allows GroupContext extensions to define simple diff formats (for example, to add, remove, and update elements in a list or items in a map).

# Conventions and Definitions

{::boilerplate bcp14-tagged}

This document uses many terms from {{!RFC9420}}. Some of these have
names which are similar or may be confusing: GroupContext, GroupContext

> extension, ExtensionType, and GroupContextExtensions proposal.

Reviewer: It might be helpful to include a brief definition list:

> * **GroupContext:** The state of the group agreed on in a given epoch
> * **GroupContext extension:** A (type, value) tuple included in the GroupContext
> * **ExtensionType:** A two-byte identifier for the type of a GroupContext extension
> * **GroupContextExtensions proposal:** A proposal to remove all of the extensions in the group's GroupContext and replace them with a new set of GroupContext extensions
> 
> Currently, the GroupContextExtensions proposal is the only mechanism defined for updating a group's GroupContext extensions.

> # GroupContext extensions modification

Reviewer: ```suggestion
# GroupContextDiff
```

In parallel with the section headings in [Section 12.1 of RFC 9420](https://rfcs.online/rfcs/rfc9420.html#name-proposals).

This document defines a new GroupContextExtensionsDiff proposal. It
is not a Safe Extension as defined in {{Section 2 of
!I-D.ietf-mls-extensions}}, because it can modify any GroupContext
extension, including those defined in {{!RFC9420}}. It is intended as a

> complete replacement for the GroupContextExtensions Proposal.

Reviewer: It doesn't necessarily replace the GCE proposal.  They can coexist.  This one is just more capable.  I think you can just delete this sentence.

> extension, including those defined in {{!RFC9420}}. It is intended as a
> complete replacement for the GroupContextExtensions Proposal.

Author: ```suggestion
extension, including those defined in {{!RFC9420}}.
```

> The GroupContextExtensionsDiff proposal does not require an UpdatePath,

Reviewer: GCE requires an UpdatePath, and this is strictly stronger than GCE, so it seems like it has to require an UpdatePath.

Author: That was listed as one of the reasons why GCE was broken AFAIK.

and may be sent by an authorized external sender.

~~~ tls
enum {
    remove(0),
    add(1),
    replace(2),
    diff(3),
    (255)
} OperationType;

uint8 DiffType;

struct {

>     ExtensionType group_context_extension;

Reviewer: ```suggestion
    ExtensionType extension_type;
```

One potential difficulty here that you should probably just note for now: I'm pretty sure that there's nothing in MLS right now that says that there can only be one extension per `ExtensionType` in the GroupContext.  If there were duplicates, this would be ambiguous.  You could fix this with something dumb like (a) defining a `noop` OperationType and (b) saying that the n-th ExtensionDiff of a given type affects the n-th entry.  That has the benefit of degrading to the normal thing in the sane case.

    OperationType operation;
    select (operation) {
        case remove:
            struct {};
        case add:
            opaque extension_data<V>;
        case replace:
            opaque extension_data<V>;
        case diff:
            opaque diff_data<V>;
    };
} ExtensionDiff;

struct {
    ExtensionDiff group_context_extension_diffs<V>;
} GroupContextExtensionsDiff;
~~~

The semantics of OperationType are as follows:

- `remove` means that the GroupContext extension was present, and will be
completely absent from the GroupContext.
- `add` means that the GroupContext extension was absent, and will be
present with the value provided in `extension_data`.
- `replace` means that the GroupContext extension was present and will
be completely replaced with the new value in `extension_data`
- `diff` means that the GroupContext extension will be modified according
to the diff format defined by that extension. (Two concrete diff formats are
defined in {{diff-formats}} which extension designers are free to use or
not use.)

> ## Processing rules

Reviewer: This doesn't need to be a subsection.  I would follow the same structure as [GCE](https://rfcs.online/rfcs/rfc9420.html#name-groupcontextextensions), "A member of the group applies a GroupContextExtensionDiff proposal with the following steps:".  That can incorporate the OperationType discussion above as well.

> 
> 
> ## Processing rules

Author: ```suggestion
```

A single GroupContextExtensionsDiff proposal can contain changes to multiple
extensions. Likewise a single Commit can include multiple
GroupContextExtensionsDiff proposals, however a single Commit MUST NOT
include multiple GroupContextExtensionsDiff proposals which refer to the
same ExtensionType. A GroupContextExtensions proposal MUST NOT appear in a
commit with any GroupContextExtensionsDiff proposal.

The motivation for allowing multiple extensions in a single
GroupContextExtensionsDiff proposal is to allow a client to modify multiple
extensions in a single "transaction". A common example of this is to add a
GroupContext extension while also adding that extension to the
`required_capabilities` GroupContext extension.

A proposal which removes a GroupContext extension that is present in the
`required_capabilities` list is invalid. Adding a required capability that is
not supported by all group members is already forbidden by {{!RFC9420}}.

> # Diff Formats

Reviewer: This section needs to specify the structure of the extensions as well, so that it's clear how the diffs are applied.  My guess is that what you want is something like:

```
opaque Opaque<V>;

struct {
    Opaque key;
    Opaque value;
} KeyValuePair;

Value VectorExtension<V>;
KeyValuePair KeyValueExtension<V>;
```
